src/Native/System.Native/pal_mount.cpp

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "../config.h"
+#include "pal_mount.h"
+#include <assert.h>
+#include <string.h>
+#include <errno.h>
+
+// Check if we should use getmntinfo or /proc/mounts
+#if HAVE_MNTINFO
+#include <sys/mount.h>
+#else
+#include <sys/statfs.h>
+#include <mntent.h>
+#define STRING_BUFFER_SIZE 8192
+#endif
+
+#if HAVE_MNTINFO
+
+int GetMountInfo(MountPointFound onFound)
+{
+    // getmntinfo returns pointers to OS-internal structs, so we don't need to worry about free'ing the object
+    int result = 0;
+    struct statfs* mounts = nullptr;
+    int count = getmntinfo(&mounts, 0);
```

Comment:
The return value of getmntinfo tells you the number of entries and the names are fixed char arrays in the struct, so it can determine all that info ahead of time; however, you have to make the function call and get back data in order to determine this. There may be a function I didn't find that just returns the count, but I couldn't find it. By going with the function pointer to C# route, it helps out the Linux implementation as well since the Linux one doesn't tell you how many at all. This seemed like the most clean method (thanks to Matt Ellis for the suggestion) for merging the two implementations...I'm happy to change it if you have an alternative that you like better :)